share/admin/share_objects.py

```diff
@@ -0,0 +1,159 @@
+from django.db.models.sql.datastructures import EmptyResultSet
+from django.utils.functional import cached_property
+
+from share.models import AbstractCreativeWork
+from share.models import Source
+from share.util import IDObfuscator
+from share.util import InvalidID
+
+
+class FuzzyPaginator(Paginator):
+
+    @cached_property
+    def count(self):
+        cursor = connections[self.object_list.db].cursor()
+
+        try:
+            cursor.execute('SELECT count_estimate(%s);', (cursor.mogrify(*self.object_list.query.sql_with_params()).decode(), ))
+        except EmptyResultSet:
+            return 0
+
+        return int(cursor.fetchone()[0])
+
+
+class SourcesInline(admin.TabularInline):
+    extra = 1
+    verbose_name = 'Source'
+    verbose_name_plural = 'Sources'
+    model = AbstractCreativeWork.sources.through
+
+    def __init__(self, parent_model, admin_site):
+        self.model = parent_model.sources.through
+        super().__init__(parent_model, admin_site)
+
+    def formfield_for_dbfield(self, db_field, **kwargs):
+        ret = super().formfield_for_dbfield(db_field, **kwargs)
+        if db_field.name == 'shareuser':
+            ret.widget.can_add_related = False
+            ret.widget.can_change_related = False
+            ret.widget.can_delete_related = False
+        return ret
+
+
+class TypedModelFilter(admin.SimpleListFilter):
+    title = 'Type'
+    parameter_name = 'type'
+
+    def lookups(self, request, model_admin):
+        return sorted([
+            (k, v._meta.verbose_name_plural.title())
+            for k, v
+            in model_admin.model._typedmodels_registry.items()
+        ])
+
+    def queryset(self, request, queryset):
+        if self.value():
+            return queryset.filter(type=self.value())
+        return queryset
+
+
+class SourcesFilter(admin.SimpleListFilter):
+    title = 'Source'
+    parameter_name = 'source'
+
+    def lookups(self, request, model_admin):
+        return sorted([
+            (source.user.username, source.long_title)
+            for source in Source.objects.select_related('user').all()
+        ], key=lambda x: x[1])
+
+    def queryset(self, request, queryset):
+        if self.value():
+            # Note: If this ever can filter on multiple sources
+            # this will require a .distinct()
+            return queryset.filter(sources__username=self.value())
+        return queryset
+
+
+class ShareObjectAdmin(admin.ModelAdmin):
+    actions = None
+    exclude = ('extra', 'same_as', 'change', )
+    inlines = (SourcesInline, )
+    list_filter = (TypedModelFilter, SourcesFilter, )
+    # Django forces order by pk desc which results in a seqscan if we add anything else
+    # ordering = ('-date_modified', )
+    readonly_fields = ('encoded_id', 'date_created', 'date_modified', )
+    show_full_result_count = False
+    paginator = FuzzyPaginator
+
+    # view_on_site doesn't actually call get_absolute_url
+    def view_on_site(self, obj):
```

Comment:
I think if get_absolute_url gave an absolute url (scheme and all), it'd work without having to specify view_on_site